### ISSUE TITLE
akka.stream.alpakka.mqtt.streaming.ActorMqttClientSession backpressure

### ISSUE THREAD

**Issue by derolf** [akka/alpakka#1581]
I have an incoming stream that gets transformed and the results is pushed to MQTT using `ActorMqttClientSession ! Command(Publish(...))`

Currently, there is no way to backpressure the incoming stream. If the stream is delivering elements very fast that results in flooding of the mailbox of the underlying actors of ActorMqttClientSession.

Interestingly, the result is that the actual TCP-writer gets only rarely scheduled:
- I see roughly one message being delivered to MQTT every 2 seconds
- The pendingLocalPublications quickly grows > 20000 elements

We need a way to throtlle/backpressure the incoming stream depending on pendingLocalPublications.

**Comment by ennru** [comment]
What you're looking for is being explored in https://github.com/akka/alpakka/pull/1577

**Comment by huntc** [comment]
Also, if you just send `Publish` commands with a QoS of 0 then there are no guarantees re. delivery. Messages could get lost. 

Where QoS > 0 then you shouldn't send messages to the same topic until you know the previous one was delivered and accepted i.e. you receive a `PubAck` event.

**Comment by derolf** [comment]
Yeah, I also thought about somehow do "manual backpressure", but have to figure out how to signal back the PubkAck event.

**Comment by huntc** [comment]
When you send a `Publish` within a `Command`, you can optionally specify an object to carry. That carry object will then appear in an associated `PubAck`'s `Event`. You can use `Promise` objects as a carry as a signaling mechanism of when to proceed.

**Comment by derolf** [comment]
@huntc Thanks for the "carry" hint. I actually made it work. The draft below pipes Protobuf-Messages through MQTT. Works.

`
import akka.actor.ActorSystem
import akka.stream.Materializer
import akka.stream.alpakka.mqtt.streaming.{scaladsl => S}
import akka.stream.alpakka.mqtt.{streaming => M}
import akka.stream.scaladsl.{Flow, Source, Tcp}
import akka.util.ByteString

import scalapb.{GeneratedMessage, GeneratedMessageCompanion, Message}

import scala.concurrent.Promise
import scala.util.Try

/**
  * A topic bound to a Protobuf message type.
  */
case class Topic[A <: GeneratedMessage with Message[A]](topic: String)(implicit val comp: GeneratedMessageCompanion[A])

/**
  * Publish a Protobuf message to a Protobuf-bound topic.
  */
case class Publish[A <: GeneratedMessage with Message[A]](topic: Topic[A], payload: A) {
  def toM(ack: () => Unit): M.Command[() => Unit] = M.Command(M.Publish(M.ControlPacketFlags.RETAIN | M.ControlPacketFlags.QoSAtLeastOnceDelivery, topic.topic, ByteString(payload.toProtoString)), ack)
}

class PbMqtt {
  val settings = M.MqttSessionSettings()

  /**
    * A flow that:
    * - consumes instance of Publish[Protobuf-Message-Type] and publishes them serialized to MQTT
    * - provides deserialized messages for the topics passed as parameters
    */
  def flow[Out](topics: Topic[_ <: Out]*)(implicit mat: Materializer, system: ActorSystem) = {
    val session = S.ActorMqttClientSession(settings)
    val connection = Flow[ByteString].via(Tcp().outgoingConnection("localhost", 1883))
    val clientSessionFlow = S.Mqtt.clientSessionFlow[() => Unit](session).join(connection)
    val commands = topics.map(topic => M.Command[() => Unit](M.Subscribe(topic.topic))).to[scala.collection.immutable.Seq]

    Flow[Publish[_]].mapAsyncUnordered(100)(publish => {
      val promise = Promise[None.type]()
      session ! publish.toM(() => promise.complete(Try(None)))
      promise.future
    }) mapConcat (_ => Nil) prepend (Source.single(M.Command[() => Unit](M.Connect("", M.ConnectFlags.CleanSession))) concat Source(commands)) via (clientSessionFlow) filter {
      case Right(M.Event(_: M.PubAck, Some(ack))) => {
        ack()
        false
      }
      case _ => true
    } collect {
      case Right(M.Event(p: M.Publish, _)) => {
        val topic = topics.find(_.topic == p.topicName)
        topic.get.comp.fromAscii(p.payload.utf8String)
      }
    }
  }
}
`
